server/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -198,15 +207,22 @@ private Object convertMapsToArrays(Map<String, Object> map) {
      * A settings that are filtered (and key is removed) with the specified prefix.
      */
     public Settings getByPrefix(String prefix) {
-        return new Settings(new FilteredMap(this.settings, (k) -> k.startsWith(prefix), prefix), secureSettings == null ? null :
-            new PrefixedSecureSettings(secureSettings, prefix, s -> s.startsWith(prefix)));
+        if (prefix.isEmpty()) {
+            return this;
+        }
+        char[] toPrefixCharArr = prefix.toCharArray();
+        toPrefixCharArr[toPrefixCharArr.length - 1]++;
```

Comment:
It took me a little bit thinking about how this works. A short comment would be appreciated, explaining this adjusts the last character of the prefix for the ending key that is non-inclusive.